File: app/src/main/java/com/veganbeauty/app/features/auth/AuthViewModel.java
package com.veganbeauty.app.features.auth;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.veganbeauty.app.data.local.entities.UserEntity;
import com.veganbeauty.app.data.repository.AuthRepository;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AuthViewModel extends ViewModel {

    private final AuthRepository authRepository;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final MutableLiveData<AuthState> _loginState = new MutableLiveData<>();
    public final LiveData<AuthState> loginState = _loginState;

    private final MutableLiveData<AuthState> _registerState = new MutableLiveData<>();
    public final LiveData<AuthState> registerState = _registerState;

    public AuthViewModel(AuthRepository authRepository) {
        this.authRepository = authRepository;
    }

    public void login(String emailOrPhone, String password) {
        if (emailOrPhone == null || emailOrPhone.trim().isEmpty() || password == null || password.trim().isEmpty()) {
            _loginState.setValue(new AuthState.Error("Vui lòng nhập đầy đủ thông tin."));
            return;
        }

        _loginState.setValue(new AuthState.Loading());
        executor.execute(() -> {
            try {
                UserEntity user = authRepository.login(emailOrPhone, password);
                if (user != null) {
                    _loginState.postValue(new AuthState.Success(user));
                } else {
                    _loginState.postValue(new AuthState.Error("Tài khoản hoặc mật khẩu không đúng."));
                }
            } catch (Exception e) {
                _loginState.postValue(new AuthState.Error("Đã xảy ra lỗi: " + e.getMessage()));
            }
        });
    }

    public void register(String fullName, String email, String phone, String password) {
        if (fullName == null || fullName.trim().isEmpty() ||
            email == null || email.trim().isEmpty() ||
            phone == null || phone.trim().isEmpty() ||
            password == null || password.trim().isEmpty()) {
            _registerState.setValue(new AuthState.Error("Vui lòng nhập đầy đủ thông tin."));
            return;
        }

        _registerState.setValue(new AuthState.Loading());
        executor.execute(() -> {
            try {
                Object resultObj = authRepository.register(fullName, email, phone, password);
                if (resultObj instanceof UserEntity) {
                    _registerState.postValue(new AuthState.Success((UserEntity) resultObj));
                } else {
                    _registerState.postValue(new AuthState.Error("Đăng ký thất bại."));
                }
            } catch (Exception e) {
                _registerState.postValue(new AuthState.Error("Đã xảy ra lỗi: " + e.getMessage()));
            }
        });
    }

    public static abstract class AuthState {
        public static class Loading extends AuthState {}
        public static class Success extends AuthState {
            private final UserEntity user;
            public Success(UserEntity user) { this.user = user; }
            public UserEntity getUser() { return user; }
        }
        public static class Error extends AuthState {
            private final String message;
            public Error(String message) { this.message = message; }
            public String getMessage() { return message; }
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/auth/AuthViewModel.kt
package com.veganbeauty.app.features.auth

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.veganbeauty.app.data.local.entities.UserEntity
import com.veganbeauty.app.data.repository.AuthRepository
import kotlinx.coroutines.launch

class AuthViewModel(private val authRepository: AuthRepository) : ViewModel() {

    private val _loginState = MutableLiveData<AuthState>()
    val loginState: LiveData<AuthState> get() = _loginState

    private val _registerState = MutableLiveData<AuthState>()
    val registerState: LiveData<AuthState> get() = _registerState

    fun login(emailOrPhone: String, password: String) {
        if (emailOrPhone.isBlank() || password.isBlank()) {
            _loginState.value = AuthState.Error("Vui lòng nhập đầy đủ thông tin.")
            return
        }

        _loginState.value = AuthState.Loading
        viewModelScope.launch {
            try {
                val user = authRepository.login(emailOrPhone, password)
                if (user != null) {
                    _loginState.value = AuthState.Success(user)
                } else {
                    _loginState.value = AuthState.Error("Tài khoản hoặc mật khẩu không đúng.")
                }
            } catch (e: Exception) {
                _loginState.value = AuthState.Error("Đã xảy ra lỗi: ${e.message}")
            }
        }
    }

    fun register(fullName: String, emailOrPhone: String, password: String) {
        if (fullName.isBlank() || emailOrPhone.isBlank() || password.isBlank()) {
            _registerState.value = AuthState.Error("Vui lòng nhập đầy đủ thông tin.")
            return
        }

        _registerState.value = AuthState.Loading
        viewModelScope.launch {
            try {
                val result = authRepository.register(fullName, emailOrPhone, password)
                if (result.isSuccess) {
                    _registerState.value = AuthState.Success(result.getOrNull()!!)
                } else {
                    _registerState.value = AuthState.Error(result.exceptionOrNull()?.message ?: "Đăng ký thất bại.")
                }
            } catch (e: Exception) {
                _registerState.value = AuthState.Error("Đã xảy ra lỗi: ${e.message}")
            }
        }
    }

    sealed class AuthState {
        object Loading : AuthState()
        data class Success(val user: UserEntity) : AuthState()
        data class Error(val message: String) : AuthState()
    }
}
